feat(vitals): Add documentation for web vitals (#2645)
getsentry/sentry-docs: src/docs/product/performance/web-vitals.mdx

---
title: "Web Vitals"
sidebar_order: 6
description: "Learn more about Web Vitals in Sentry and how they give you a better understanding of your application's health."
---

Frontend transactions can additionally contain web vital information. With this view, you'll be able to examine a summary of these web vitals.

> A summary is generated for each of the following web vitals.

Reviewer: This felt redundant. If you want to explicitly talk about the table. You could write something like, "The following table provides details about the web vitals."

| Web Vital                                                 | Availability                  | Failure Threshold |
| --------------------------------------------------------- | ----------------------------- | ----------------- |
| [First Paint](#first-paint-fp)                            | Chromium                      | 3 seconds         |
| [First Contentful Paint](#first-contentful-paint-fcp)     | Chromium                      | 3 seconds         |
| [Largest Contentful Paint](#largest-contentful-paint-lcp) | Chromium                      | 4 seconds         |
| [First Input Delay](#first-input-delay-fid)               | Chromium, Firefox, and Safari | 300 milliseconds  |
| [Cumulative Layout Shift](#cumulative-layout-shift-cls)   | Chromium                      | 0.25              |

## Baseline Value

> The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

Author: The 75th percentile and failure threshold here differ from the p75 thrshold and failure rate in the docs you linked.

1. The 75th percentile is referring to the generic 75th percentile for the web vital values, not the `transaction.duration` like in those docs
2. The failure threshold here is referring to the values listed in the above table that corresponds to each web vital, not the percentage of transactions with a failed status on it.

> The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

Reviewer: Ah, I see. Thanks for the clarification, Tony.

> The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

Author: Is the wording here perhaps confusing and caused you to misunderstand?

> The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

Reviewer: I think some of the terminologies here might get confused with Performance Metrics, which uses similar vocabulary. Maybe we should explicitly call out that they're different?

> The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

Reviewer: Maybe add a Note tag and write something like:
Web Vital terminologies such as "75th percentile" and "failure threshold" are different from similar terminology used for Performance Metrics.

## Distribution Histogram

A histogram visualization showing the distribution of the data can help you identify and diagnose frontend performance problems by revealing anomalies.

By default, outliers will be excluded from the histograms to provide a more informative view of these vitals. Outliers are determined by using the [upper outer fence](https://www.itl.nist.gov/div898/handbook/prc/section1/prc16.htm) as the upper bound, and any data points above the upper bound are deemed to be an outlier. If you wish to see all of the data available, you can still do so.

> If you see a region of interest on any of the histograms, click and drag over the area to zoom in for a more detailed view. You may also want to see more information related to the transactions in the histograms. Click "Open in Discover" beneath the web vital of choice to build a custom query to investigate further. For more details, see the full documentation for the Discover [Query Builder](/product/discover-queries/query-builder/).

Author: I'm not sure about the change to say
> all of the data available
here as the sentence now seems to suggest that you click and drag to see all of the data. This is not the case as click and dragging is only for zooming in.

> If you see a region of interest on any of the histograms, click and drag over the area to zoom in for a more detailed view. You may also want to see more information related to the transactions in the histograms. Click "Open in Discover" beneath the web vital of choice to build a custom query to investigate further. For more details, see the full documentation for the Discover [Query Builder](/product/discover-queries/query-builder/).

Reviewer: In that case, I agree. Let's reword it. I think a previous version mentioned how you could see all the data available but didn't explicitly describe how. Is that still true? Is there a way to see all the data available? If so, I'd break this up into two separate sentences. One sentence for how to see all the available data and another sentence for zooming in.

> If you see a region of interest on any of the histograms, click and drag over the area to zoom in for a more detailed view. You may also want to see more information related to the transactions in the histograms. Click "Open in Discover" beneath the web vital of choice to build a custom query to investigate further. For more details, see the full documentation for the Discover [Query Builder](/product/discover-queries/query-builder/).

Author: It is possible to see all the data available. There is a dropdown menu on the page that allows you to do this. I'll update this to talk about that as well.

![image](https://user-images.githubusercontent.com/10239353/98305780-83af7d80-1f90-11eb-9a15-01831ef11255.png)

> If you see a region of interest on any of the histograms, click and drag over the area to zoom in for a more detailed view. You may also want to see more information related to the transactions in the histograms. Click "Open in Discover" beneath the web vital of choice to build a custom query to investigate further. For more details, see the full documentation for the Discover [Query Builder](/product/discover-queries/query-builder/).

Reviewer: Thanks, Tony.

### Largest Contentful Paint (LCP)

Largest Contentful Paint measures the render time for the largest content to appear in the viewport. This may be in any form from the document object model (DOM) such as images, svgs, text blocks, etc. It is the largest pixel area in the viewport, thus most visually defining. It helps developers understand how long it takes before the user perceives the main content on the page.

### First Input Delay (FID)

First Input Delay measures the response time when the user tries to interact with the viewport. Actions may include clicking a button, link or other custom Javascript controller. It is key in helping the user determine if a page is usable or not.

### Cumulative Layout Shift (CLS)

Cumulative Layout Shift is the sum of individual layout shift scores for every unexpected element shift that occurs during the rendering process. Imagine navigating to an article and trying to click a link before the page finishes loading. Before your cursor even gets there, the link may have shifted down due to an image rendering. Rather than using duration for this web vital, the score represents how disruptive and visually unstable the shifts are.

> Each layout shift score is calculated using an impact and distance fraction where impact fraction is the total visible area that the element affects between the two rendered frames and distance fraction measures the distance it has moved relative to the viewport.

Reviewer: ```suggestion
Each layout shift score is calculated using an impact and distance fraction. The impact fraction is the total visible area that the element affects between the two rendered frames. The distance fraction measures the distance it has moved relative to the viewport.
```

Layout shift score = Impact Fraction * Distance Fraction

> Let’s take a look at the example above which has one unstable element - the body text. The impact fraction is roughly 50% of the page and moves the body text down by 20%. The layout shift score is 0.1, the product of 0.5*0.2. Thus, CLS is 0.1.

Reviewer: I think we're missing this exact example above.

Author: Yes, I haven't added the necessary screenshots yet. I'm tracking them in the TODO in the description above and will be checking them off as they're added in 😄

Reviewer: Great. Thanks, Tony! 🌻

---
title: "Web Vitals"
sidebar_order: 6

> description: "Learn more about Web Vitals in Sentry and how they give you a better understanding of your application's overall health."

Reviewer 2: I feel like we don't need `overall` here.

```suggestion
description: "Learn more about Web Vitals in Sentry and how they give you a better understanding of your application's health."
```

---

> Some transactions will additionally contain web vital information. With this view, you'll be able to examine a summary of these web vitals.

Reviewer 2: Should we call out that it's frontend, pageload transactions and only captured if the user is using Chrome?

- [Largest Contentful Paint (LCP)](#largest-contentful-paint-lcp)
- [First Input Delay (FID)](#first-input-delay-fid)
- [Cumulative Layout Shift (CLS)](#cumulative-layout-shift-cls)

## Baseline Value

The baseline value for each web vital is the 75th percentile of the observed data points. It is compared against a failure threshold to give a pass or fail for each web vital.

### Thresholds

| Web Vital                                                 | Threshold        |
| --------------------------------------------------------- | ---------------- |
| [First Paint](#first-paint-fp)                            | 3 seconds        |
| [First Contentful Paint](#first-contentful-paint-fcp)     | 3 seconds        |
| [Largest Contentful Paint](#largest-contentful-paint-lcp) | 4 seconds        |
| [First Input Delay](#first-input-delay-fid)               | 300 milliseconds |
| [Cumulative Layout Shift](#cumulative-layout-shift-cls)   | 0.25             |

## Distribution Histogram

A histogram visualization showing the distribution of the data can help you identify and diagnose frontend performance problems by revealing anomalies.

> By default, outliers will be excluded from the histograms to provide a more informative view of these vitals. Outliers are determined by using the [upper outer fence](https://www.itl.nist.gov/div898/handbook/prc/section1/prc16.htm) as the upper bound, and any data points above the upper bound are deemed to be an outlier. If you wish to see all of the data available, you can still do so.

Author: Not sure about this link as a reference. Can't find another reference right now.

Reviewer 2: https://www.thoughtco.com/what-are-inner-and-outer-fences-3126374 maybe?

Author: Using this wikipedia link instead https://en.wikipedia.org/wiki/Outlier#Tukey's_fences

---
title: "Web Vitals"
sidebar_order: 6
description: "Learn more about Web Vitals in Sentry and how they give you a better understanding of your application's health."
---

> Frontend transactions can additionally contain web vital information. With this view, you'll be able to examine a summary of the web vitals.

Reviewer 3: ```suggestion
Frontend transactions may contain web vitals information if the browser provides them. With this view, you'll be able to examine a summary of the web vitals.
```

> Frontend transactions can additionally contain web vital information. With this view, you'll be able to examine a summary of the web vitals.

Reviewer 3: We'll want a link to https://web.dev/vitals/ somewhere in this paragraph.

Author: Added in 680c827

| Web Vital                                                 | Availability                  | Failure Threshold |
| --------------------------------------------------------- | ----------------------------- | ----------------- |
| [First Paint](#first-paint-fp)                            | Chromium                      | 3 seconds         |
| [First Contentful Paint](#first-contentful-paint-fcp)     | Chromium                      | 3 seconds         |
| [Largest Contentful Paint](#largest-contentful-paint-lcp) | Chromium                      | 4 seconds         |
| [First Input Delay](#first-input-delay-fid)               | Chromium, Firefox, and Safari | 300 milliseconds  |

> | [Cumulative Layout Shift](#cumulative-layout-shift-cls)   | Chromium                      | 0.25              |

Reviewer 3: We'll need to add TTFB here.

By default, outliers will be excluded from the histograms to provide a more informative view of these vitals. Outliers are determined using the [upper outer fence](https://www.itl.nist.gov/div898/handbook/prc/section1/prc16.htm) as the upper bound, and any data points above the upper bound are deemed an outlier.

If you notice a region of interest on any of the histograms, click and drag over the area to zoom in for a more detailed view. You may also want to see more information related to the transactions in the histograms. Click "Open in Discover" beneath the web vital of choice to build a custom query for further investigation. For more details, see the full documentation for the Discover [Query Builder](/product/discover-queries/query-builder/).

If you wish to see all of the data available, open the dropdown and click "View All". You will likely see extreme outliers when you click "View All". You can click and drag over an area to zoom in for a more detailed view.

## Vitals

Vitals measure render time, response time, and the sum of shifting rendering processes. Each data point provides insights about the overall performance of your application.

### First Paint (FP)

> First Paint (FP) measures the amount of time the first pixel takes to appear in the viewport, rendering any visual change from what was previously displayed. This may be in any form from the document object model (DOM), such as background color, canvas, or image. FP helps developers understand if anything unexpected is happening to render the page.

Reviewer 3: ```suggestion
First Paint (FP) measures the amount of time the first pixel takes to appear in the viewport, rendering any visual change from what was previously displayed. This may be in any form from the document object model (DOM), such as non-default background, canvas, or image. FP helps developers understand if anything unexpected is happening to render the page.
```

reference: https://w3c.github.io/paint-timing/#first-paint

Author: Going to leave this as just

> background color

### First Contentful Paint (FCP)

> First Contentful Paint (FCP) measures the time for the first content to render in the viewport. This may be in any form from the document object model (DOM), such as images, SVGs, or text blocks. FCP frequently overlaps with First Paint (FP). FCP helps developers understand how long it takes before the user sees any content change on the page.

Reviewer 3: I highly recommend we should link to https://web.dev/fcp/ for more in-depth information. Likewise for other web vital sections.

Author: Added in 680c827
